jupyter_collaboration/rooms.py

```diff
@@ -0,0 +1,154 @@
+import asyncio
+from logging import Logger
+from typing import Any, Optional
+
+from jupyter_ydoc import ydocs as YDOCS
+from ypy_websocket.websocket_server import YRoom
+from ypy_websocket.ystore import BaseYStore, YDocNotFound
+
+from .loaders import FileLoader
+
+YFILE = YDOCS["file"]
+
+
+class DocumentRoom(YRoom):
+    """A Y room for a possibly stored document (e.g. a notebook)."""
+
+    def __init__(
+        self,
+        room_id: str,
+        file_format: str,
+        file_type: str,
+        file: FileLoader,
+        ystore: Optional[BaseYStore],
+        log: Optional[Logger],
+        save_delay: Optional[int] = None,
+    ):
+        super().__init__(ready=False, ystore=ystore, log=log)
+
+        self._room_id: str = room_id
+        self._file_format: str = file_format
+        self._file_type: str = file_type
+        self._file: FileLoader = file
+        self._document = YDOCS.get(self._file_type, YFILE)(self.ydoc)
+
+        self._save_delay = save_delay
+
+        self._lock = asyncio.Lock()
+        self._cleaner: Optional[asyncio.Task] = None
+        self._saving_document: Optional[asyncio.Task] = None
+
+        # Listen for document changes
+        self._document.observe(self._on_document_change)
+        self._file.observe(self.room_id, self._on_content_change)
+
+    @property
+    def room_id(self) -> str:
+        return self._room_id
+
+    @property
+    def cleaner(self) -> Optional[asyncio.Task]:
+        return self._cleaner
+
+    @cleaner.setter
+    def cleaner(self, value: asyncio.Task) -> None:
+        self._cleaner = value
+
+    def _clean(self) -> None:
+        super()._clean()
+        # TODO: Should we cancel or wait ?
```

Comment:
I would say that we should wait, but that needs changing some methods to async in `ypy-websocket`.

Reply:
Then I think we should leave it like this for now.

This is a rare case. Every time the user stops typing we save the content with 1-second delay.